### ISSUE TITLE
Semantics of `Free.suspend` changed in 0.7.0

### ISSUE THREAD

**Issue by mpilquist** [typelevel/cats#1277]
I'm not sure if this is an issue or not but we changed semantics of Free.suspend in 0.7:

```scala
scala> val x = Free.suspend { println("Computing"); Free.pure(()) }
x: cats.free.Free[S,Unit] = Free(...)

scala> x.step
Computing
res0: cats.free.Free[S,Unit] = Free(...)
````

Whereas in 0.6.1:

```scala
@ val x = Free.suspend[List,Unit] { println("Computing"); Free.pure(()) }
x: Free[List, Unit] = Gosub(Suspend(List(())),<function1>)
@ x.step
res3: Free[List, Unit] = Gosub(Suspend(List(())),<function1>)
```

This is because step [eagerly evaluates `FlatMapped(Pure(x), f)` when stepping](https://github.com/typelevel/cats/blob/97fc7625923076180a5a2dd7f03798882c85c26d/free/src/main/scala/cats/free/Free.scala#L39), which used to be safe in presence of `suspend`, b/c [`suspend` included a call to `liftF` and hence, a `Suspend` constructor](https://github.com/typelevel/cats/blob/v0.6.1/free/src/main/scala/cats/free/Free.scala#L28).

I was bitten by this same bug in FS2, where suspend has always been defined as `pure(()).flatMap { _ => fa }`. In FS2, I removed the eager evaluation of `FlatMapped(Pure(x), f)` b/c it impacts semantics of effect capture and that's half of what FS2 is about. Not sure what the right thing for Cats is though...

Some options:

(1) Restore old definition, which means suspend requires an `Applicative[F]` for pure
(2) Define both constructors, one which is guaranteed not to be evaluated while stepping (names TBD)
(3) Remove the eager application of `FlatMapped(Pure(x), f)` from step
(4) Ignore this since it is only visible in presence of side effects

Does anyone know the genesis of the eager evaluation in `step` (maybe @d6)? As far as I can tell, this is a performance enhancement to avoid some trampolining.

There was [some discussion on Gitter](https://gitter.im/typelevel/cats?at=57a8fa6d46610f173949bf46) that may be interesting. 

/cc @tpolecat @adelbertc

**Comment by mikejcurry** [comment]
@mpilquist - I think you meant @non as opposed to `d6`? - apologies if I got that wrong.

**Comment by mpilquist** [comment]
@mikejcurry Indeed, thanks, and my apologies to `d6`. :)

**Comment by ceedubs** [comment]
My initial reaction is that Cats should be consistent with the approach that it has taken in other areas, which is that the execution semantics are side effects are not guaranteed if you are not tracking your side effects explicitly in an `IO`/`Task`-like type. Making this assumption is something that allows us to make certain optimizations (like this seems to be). I think that for the most part if people are relying on subtleties of side effect execution without explicitly tracking the effects in a type, it's probably going to bite them sooner or later.

Having said that, I don't feel _super_ strongly about this, so I'm happy to go along with whatever others decide.

One question: an `IO` type might use `Free` under the covers. Does the current behavior hinder the ability to do that? Or can the `IO` type capture its own thunk in such a way that this isn't a problem?

**Comment by mpilquist** [comment]
> One question: an IO type might use Free under the covers. Does the current behavior hinder the ability to do that? Or can the IO type capture its own thunk in such a way that this isn't a problem?

This is sort of what happened with FS2. A stream is compiled to a `Free[Task,A]` which is then interpreted in to a `Task[A]`. Our bug was that the resulting task was safe to run once but failed to run any subsequent times, due to an allocation that was wrapped in `Free.suspend` that was optimized away by `step`.

**Comment by tpolecat** [comment]
As stated previously in gitter, I vote for (4) above. Effect capture is a special operation that must be engineered, it's not something you can get on the sly by trying to reason about evaluation semantics. Arguments to `capture` are side-effects and cannot be factored out; equational reasoning doesn't work here. If Scala had a a `<flash>` tag that's where you would use it.

I need to put that on hotkey, or replace myself with a small shell script.

**Comment by adelbertc** [comment]
Having written a retry library that was predicated on being able to re-run effects, I agree with @tpolecat . When trying to parameterize my retries with `F[_]: scalaz.Monad` and taking advantage of by-name `point` it felt wrong since it made sense for certain data types (`IO`, `Task`, `Name`), but not for anything that evaluated it right away anyways (like `Option`). This raises the question if it's possible to have a separate type class that abstracts over this, something like doobie's `Capture` but being lawless it seems quite sketchy. Having it be a special operation on specific data types seems like the way to go.

This all just boils down to the argument of "IO: There can only be one" though the reality of the Scala world we live in has many, hence the sketchiness of `Capture`.

tl;dr I vote (4)

**Comment by non** [comment]
I am largely sympathetic to (4), so I think that's where my vote falls.

If we can come up with clear principles for when `Free` should (or shouldn't) evaluate its values I'm open to changing my mind, but for now I think I'm erring on the side of performance.

**Comment by mpilquist** [comment]
Closing this with no changes since that's the overwhelming preference.
